File: objc/Adapter/Adapter/ThunderboltAdapter.m
#import "ThunderboltAdapter.h"

@implementation ThunderboltAdapter {
    id<Connection> connection;
}

-(instancetype)initWithConnection:(id<Connection>)conn {
    if(self = [super init]) {
        connection = conn;
    }
    return self;
}

-(void)sendData:(NSString*)data {
    if([connection isKindOfClass:[EthernetConnection class]]) {
        for(int i=0; i < data.length; i += 10) {
            long len = (i + 10) > data.length ? data.length - i : 10;
            [connection sendPacket:[data substringWithRange:NSMakeRange(i, len)]];
        }
        return;
    }
    
    if([connection isKindOfClass:[MobileDataConnection class]]) {
        for(int i=0; i < data.length; i += 5) {
            long len = (i + 5) > data.length ? data.length - i : 5;
            [connection sendPacket:[data substringWithRange:NSMakeRange(i, len)]];
        }
        return;
    }
}

@end

File: swift/Adapter/Adapter/ThunderboltAdapter.swift
//
//  ThunderboltAdapter.swift
//  Adapter
//
//  Created by KAZI MAHBUBUR RAHMAN on 24/3/18.
//  Copyright © 2018 KAZI MAHBUBUR RAHMAN. All rights reserved.
//

import Foundation

class ThunderboltAdapter {
    var connection: Connection
    
    init(_ connection: Connection) {
        self.connection = connection
    }
    
    func sendData(_ data: String) {
        
        if connection is EthernetConnection {
            let sequence = stride(from: 0, to: data.count, by: 10)
            for i in sequence {
                let start = data.index(data.startIndex, offsetBy: i)
                let end = i + 10 >= data.count ?
                    data.index(data.endIndex, offsetBy: 0) :
                    data.index(data.startIndex, offsetBy: i+10)
                connection.sendPacket(String(data[start..<end]))
            }
            return
        }
        
        if connection is MobileDataConnection {
            let sequence = stride(from: 0, to: data.count, by: 5)
            for i in sequence {
                let start = data.index(data.startIndex, offsetBy: i)
                let end = i + 5 >= data.count ?
                    data.index(data.endIndex, offsetBy: 0) :
                    data.index(data.startIndex, offsetBy: i+5)
                connection.sendPacket(String(data[start..<end]))
            }
            return
        }
    }
}
